hail/src/main/scala/is/hail/expr/ir/Emit.scala

```diff
@@ -2430,6 +2432,71 @@ private class Emit(
           }
         }
 
+      case x@NDArraySlice(child, slicesIR) =>
+        val childEmitter = deforest(child)
+
+        val slicest = emit(slicesIR, env, resultRegion, None)
+        val slicesValueAddress = mb.newField[Long]
+        val slicesTuple = new CodePTuple(coerce[PTuple](slicesIR.pType), region, slicesValueAddress)
+
+        val codeSlices = ArrayBuffer[(Code[Long], Code[Long], Code[Long])]()
+        val codeRefMap = mutable.Map[Int, Code[Long]]()
+
+        var missingSliceElements = const(false)
+
+        coerce[PTuple](slicesIR.pType).types.zipWithIndex.foreach { case (sliceOrIndex, dim) =>
```

Comment:
I think this will read more clearly with some helper functions:
```scala
class CodePTuple(...) {
  ...
  def withTypes = withTypesAndIndex.map(x => x._1 -> x._2)

  def withTypesAndIndex = (0 until pType.nFields).map(i => (pType.types(i), apply(i), i))

  def missingness = (0 until pType.nFields).map(isMissing _)

  // we can write this out for sizes 1 through 3
  def values[T, U, V] = {
    assert pType.nFields == 3
    (apply[T](0), apply[U](1), apply[V](2))
  }
}
```
and:
```scala
slicesTuple.withTypesAndIndex {
  case (p: PTuple, _slice, _) =>
    val slice = new CodePTuple(p, region, _slice)
    missingElements ||= slice.missingness.fold(_ || _)
    codeSlices += slice.values[Long, Long, Long]
  case (_: PInt64, index, dim) =>
    codRefMap(dim) = index
```

Reply:
Actually might be more clear as two separate steps:
```scala
val slicers = slicesTuple.withTypes.collect {
    case (t: PTuple, slice) => new CodePTuple(t, region, slice) }
val codeSlices = slicers.map(_.values[Long, Long, Long])

val indexers = slicesTuple.withTypesAndIndex.collect {
    case (_: PInt64, indexer, i) => i => coerce[Long](indexer) }
val codeRefMap = indexers.toMap

val anyMissingness = slicesTuple.missingness.fold(_ || _) ||
    slicers.map(_.missingness.fold(_ || _)).fold(_ || _)
```

Reply:
I prefer a more descriptive name for `codeRefMap`, maybe something like `indexer`? `selector`? `nonSlicer`?